File: backend/src/services/releaseService.js
const fs = require('fs');
const path = require('path');

const RELEASE_FILE = path.join(__dirname, '..', 'config', 'release.sample.json');

function compareVersions(a, b) {
  const av = String(a).split('.').map((n) => Number(n) || 0);
  const bv = String(b).split('.').map((n) => Number(n) || 0);
  const len = Math.max(av.length, bv.length);

  for (let i = 0; i < len; i++) {
    const ai = av[i] ?? 0;
    const bi = bv[i] ?? 0;
    if (ai > bi) return 1;
    if (ai < bi) return -1;
  }
  return 0;
}

function getReleaseInfo() {
  const rawText = fs.readFileSync(RELEASE_FILE, 'utf-8');
  return JSON.parse(rawText);
}

function checkRelease({ currentVersion, platform }) {
  const release = getReleaseInfo();
  const latestVersion = String(release.latestVersion || '0.0.0');
  const cmp = compareVersions(String(currentVersion || '0.0.0'), latestVersion);

  return {
    hasNewVersion: cmp < 0,
    latestVersion,
    currentVersion: String(currentVersion || '0.0.0'),
    updateUrl: platform === 'ios' ? release.iosStoreUrl : release.androidStoreUrl,
    releaseNotes: Array.isArray(release.releaseNotes) ? release.releaseNotes.join('\n') : '',
  };
}

module.exports = {
  checkRelease,
};

File: backend/python_import_service/services/release_service.py
from pathlib import Path
import json


RELEASE_FILE = Path(__file__).resolve().parent.parent / "data" / "release.sample.json"


def compare_versions(a: str, b: str) -> int:
    a_parts = [int(part) if part.isdigit() else 0 for part in str(a).split(".")]
    b_parts = [int(part) if part.isdigit() else 0 for part in str(b).split(".")]
    max_len = max(len(a_parts), len(b_parts))
    for index in range(max_len):
        left = a_parts[index] if index < len(a_parts) else 0
        right = b_parts[index] if index < len(b_parts) else 0
        if left > right:
            return 1
        if left < right:
            return -1
    return 0


def load_release_info() -> dict:
    if not RELEASE_FILE.exists():
        return {
            "latestVersion": "0.0.0",
            "androidStoreUrl": "",
            "iosStoreUrl": "",
            "releaseNotes": [],
        }
    return json.loads(RELEASE_FILE.read_text(encoding="utf-8"))


def check_release(current_version: str, platform: str) -> dict:
    release = load_release_info()
    latest_version = str(release.get("latestVersion", "0.0.0"))
    cmp = compare_versions(current_version, latest_version)
    return {
        "hasNewVersion": cmp < 0,
        "latestVersion": latest_version,
        "currentVersion": current_version,
        "updateUrl": release.get("iosStoreUrl", "") if platform == "ios" else release.get("androidStoreUrl", ""),
        "releaseNotes": "\n".join(release.get("releaseNotes", [])),
    }